node/src/validator/router.rs

```diff
@@ -38,6 +46,77 @@ impl<N: Network, C: ConsensusStorage<N>> P2P for Validator<N, C> {
     }
 }
 
+impl<N: Network, C: ConsensusStorage<N>> Validator<N, C> {
+    /// An extended handshake dedicated to consensus committee members.
+    async fn committee_handshake(
+        &self,
+        peer_ip: SocketAddr,
+        mut framed: Framed<&mut TcpStream, MessageCodec<N>>,
+    ) -> io::Result<()> {
+        // Establish quorum with other validators:
+        //
+        // 1. Sign and send the node's pub key.
+        // 2. Receive and verify peer's signed pub key.
+        // 3. Insert into connected_committee_members.
+        // 4. If quorum threshold is reached, start the bft.
+
+        // 1.
+        // BFT must be set here.
+        // TODO: we should probably use something else than the public key, potentially interactive, since this could
+        // be copied and reused by a malicious validator.
+        let public_key = self.primary_keypair.public();
+        let signature = self
+            .primary_keypair
+            .private()
+            .sign_bytes(public_key.to_bytes().as_slice(), &mut rand::thread_rng())
+            .unwrap();
+
+        let last_executed_sub_dag_index =
+            if let Some(bft) = self.bft.get() { bft.state.last_executed_sub_dag_index().await } else { 0 };
+        let message = Message::ConsensusId(Box::new(ConsensusId {
+            public_key: public_key.clone(),
+            signature,
+            last_executed_sub_dag_index,
+        }));
+        framed.send(message).await?;
+
+        // 2.
+        let consensus_id = match framed.try_next().await? {
+            Some(Message::ConsensusId(data)) => data,
+            _ => return Err(error(format!("'{peer_ip}' did not send a 'ConsensusId' message"))),
+        };
+
+        // Check the advertised public key exists in the committee.
+        if !self.committee.keys().contains(&&consensus_id.public_key) {
+            return Err(error(format!("'{peer_ip}' is not part of the committee")));
+        }
+
+        // Check the signature.
+        // TODO: again, the signed message should probably be something we send to the peer, not
+        // their public key.
+        if !consensus_id.signature.verify_bytes(&consensus_id.public_key, &consensus_id.public_key.to_bytes()) {
+            return Err(error(format!("'{peer_ip}' couldn't verify their identity")));
+        }
+
+        // 3.
+        // Track the committee member.
+        // TODO: in future we could error here if it already exists in the collection but that
+        // logic is probably best implemented when dynamic committees are being considered.
+        self.connected_committee_members.write().insert(peer_ip, consensus_id.public_key);
```

Comment:
^Read the TODO first.

Need to dedup the public key. Otherwise, a compromised validator would be able to sabotage the actual validator.